kernel/mempool.c

```diff
@@ -146,8 +147,15 @@ void *k_mem_pool_malloc(struct k_mem_pool *pool, size_t size)
 		return NULL;
 	}
 
+	/*
+	 * MISRA-C 21.15 requires the pointer arguments for memcpy be
+	 * compatible types. Unfortunately just cast block.id inside the function
+	 * is not enough.
+	 */
+	aux = &block.id;
+
 	/* save the block descriptor info at the start of the actual block */
-	(void)memcpy(block.data, &block.id, sizeof(struct k_mem_block_id));
+	(void)memcpy(block.data, aux, sizeof(struct k_mem_block_id));
```

Comment:
If we're going to have to do this, let's at least do it with struct assignment, which is shorter, no less typesafe and no less clear (though less idiomatic, I guess), and avoids needing the MISRA comment explaining why it has to be done this way.

`
    *(struct k_mem_block_id *)block.data = block.id;
`

Reply:
I'm afraid I've tested it and this violates others MISRA-C rules regarding essential types. I'll double check anyway.